### PR TITLE
Add a `linear_fake_merge` option (defaulting to True)

### PR DESCRIPTION
**Pull request by cgwalters**
One of my projects (projectatomic/rpm-ostree) has a protected master
branch.  I was rather horrified to realize that homu force-pushes to
master by default.  Admittedly, I understand this is the only way to
make github think the PR was merged (as opposed to just closed).

But in my mind, that's a deficiency of github - it should have an API
to allow us to "close merged with this commit".

Anyways, add an option so that those who prefer closed over force
pushing can do that today.

We at least retitle the commit so that it's more clear it was merged.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/homu/19)

### CODE REVIEW

File: homu/main.py [hunk 1]
```diff
@@ -216,7 +216,12 @@ def refresh(self):
         self.body = issue.body
 
     def fake_merge(self, repo_cfg):
```
Comment (Manishearth) [inline homu/main.py]: Isn't fake_merge only used when exempting builds?

Comment (cgwalters) [inline homu/main.py]: What do you mean by "exempting"?  `git grep exempting` doesn't know.

**Comment by cgwalters** [comment]
It's pretty tempting to flip the default for this to on...having a constantly force-pushed master branch seems like something we shouldn't encourage by default.  (Only affecting users who use an already non-default `linear = true` option)

File: cfg.sample.toml [hunk 2]
```diff
@@ -44,6 +44,13 @@ try_users = []
 # Auto-squash commits. Requires the local Git command
 #autosquash = true
 
+# Whether homu should push a temporary merge commit that causes
```
Comment (Manishearth) [inline cfg.sample.toml]: To be clear, this only applies if `linear=true`, right? If so, that should be mentioned here.

File: homu/main.py (hunk 1 above)
Comment (cgwalters) [inline homu/main.py]: Oh sorry, `git grep exempt` showed me.  The answer is this is also called by `report_build_res()`.

File: cfg.sample.toml (hunk 2 above)
Comment (cgwalters) [inline cfg.sample.toml]: Done in a fixup! :arrow_down:

**Comment by cgwalters** [comment]
Related to this, I found https://github.com/isaacs/github/issues/2 which is a top google result for this.  Hopefully github does something eventually to support non-merge-button workflows better.

**Comment by cgwalters** [comment]
Update: Github [introduced squashing](https://github.com/blog/2141-squash-your-commits).  However, I still like Homu's flow better since:

- It does the test-before-merge
- It supports `fixup!` (though this seems like an obvious feature for github to add now)
- Related to test-before-merge, I like that I can just drop an r+ in a comment even *before* any PR tests come in and know if the tests pass, it will be merged and I won't have to look at the PR again.  Nicer for smaller PRs.

**Comment by cgwalters** [comment]
I'm wondering if there any Homu users with `linear=true` who would really be okay with the force pushing. 

It seems like the current high profile users (Rust/Servo) don't use `linear=true`, and for that matter homu.io doesn't seem to allow configuring it.

Some git blame archaeology leads me to https://github.com/barosl/homu/issues/58 - so if it was just Servo that wanted it but didn't end up using the implementation, I'm inclined to "take over" the feature here and fix the force pushing, since again I think it's a dangerous and bad idea.

**Comment by Ms2ger** [comment]
Getting rid of force pushing seems like a good idea.

**Comment by cgwalters** [comment]
Okay, reworked the commit :arrow_up: to drop the previous behavior entirely.  Ideally we'd have a mechanism to communicate with Homu users, but in my searches there aren't many who are using `linear`, and the ones who do (people I know) would all want this new behavior.

**Comment by cgwalters** [comment]
Friendly ping - I'd like to promote use of Homu more but it's problematic when some of the features I'm selling aren't in git master.

File: homu/main.py [hunk 3]
```diff
@@ -216,30 +216,17 @@ def refresh(self):
...
-                self.get_issue().close()
+        if not repo_cfg.get('linear', False) or repo_cfg.get('autosquash', False):
+            return
 
-            utils.retry_until(inner, fail, self)
+        issue = self.get_issue()
+        title = issue.title
+        # We tell github to close the PR via the commit message, but it doesn't know that
+        # constitutes a merge.  Edit the title so that it's clearer.
+        merged_prefix = '[merged] '
```
Comment (Manishearth) [inline homu/main.py]: So github doesn't actually merge this? Could we add an option  to toggle this behavior?

Comment (Manishearth) [inline homu/main.py]: Actually nvm, I missed one of your comments above.

**Comment by Manishearth** [comment]
Sorry about the delay; I'm extremely busy these days, but in a month I should be able to spend more time on homu.

File: homu/main.py (hunk 3 above)
Comment (Manishearth) [inline homu/main.py]: Perhaps this should also close the PR?

Comment (cgwalters) [inline homu/main.py]: See the hunk below, that's what the `Closes: ` verb does.

BTW,  you can see real world examples of this:
https://github.com/ostreedev/ostree/pull/278
https://github.com/projectatomic/rpm-ostree/pull/270

Comment (Manishearth) [inline homu/main.py]: Ah, I didn't realize that PRs could be closed with closes. Nice

**Pull request merged**